### DISCUSSION TITLE
"Why Low-Precision Transformer Training Fails: An Analysis on Flash Attention"

### DISCUSSION THREAD

**Discussion by AustenMan** [Dao-AILab/flash-attention#1931] [General]
I'm curious if the authors have seen this paper, and whether a toggle could be added for this stability patch they proposed?

https://arxiv.org/pdf/2510.04212

**Comment by tridao** [comment]
Thanks for sharing, I just skimmed it. Quite interesting and I'll need to read more carefully later.
The proposed method (Sec 4) doesn't look too hard to implement. We welcome contributions. 
To modify the softmax computation: 
- Cute-DSL version: https://github.com/Dao-AILab/flash-attention/blob/35384ecdf5461a79cf39d5c547185a4c89b91b5d/flash_attn/cute/softmax.py#L51
- CUDA C++ version: https://github.com/Dao-AILab/flash-attention/blob/35384ecdf5461a79cf39d5c547185a4c89b91b5d/csrc/flash_attn/src/softmax.h#L137

**Comment by tridao** [comment]
Briefly looking at the code, looks like the authors are using a Pytorch reimplementation of FA, which I worry may not have the same numerical behavior as the FA implementation here. E.g. here we keep accumulator in fp32, use fused multiply add (fma) instead of separate mul and add, etc. Those details all affect numerical error.
https://github.com/ucker/why-low-precision-training-fails/blob/67811bd77435a5bcd6a5cb56ccd8a7be8d328a6b/attention.py#L22
